### DISCUSSION TITLE
How to update Wintercms ?

### DISCUSSION THREAD

**Discussion by TrenchTown** [wintercms/winter#279] [Q&A] [answered]
Hi, all
I have version 1.1.3 installed, how do I upgrade to version 1.1.4 ?

I tried it in the admin panel, it gives a message that there are no updated versions

**Comment by bennothommo** [comment, marked as answer]
@TrenchTown you'll need to use Composer at this stage. Run `composer update` inside the project folder.

**Reply by TrenchTown** [reply]
Thanks.
Is the function planned to be updated via the admin panel?

**Reply by bennothommo** [reply]
Yep, when our Marketplace is online.

**Reply by WebVPF** [reply]
When Marketplace is online?

**Reply by LukeTowers** [reply]
Hopefully September or October.

**Reply by ad-sharky** [reply]
I'm sorry, how to update from 1.1.9 to 1.2.0, composer update unfortunately does not give the desired results. php artisan winter:update
Updating Winter...
No new updates found

**Reply by ad-sharky** [reply]
php artisan winter:version
*** Detecting Winter CMS build...
*** Detected a modified version of Winter CMS build 1.1.9.

**Reply by AIC-BV** [reply]
@ad-sharky follow the upgrade guide: https://github.com/wintercms/meta/blob/master/l9-upgrade-notes.md
There are several breaking changes

**Reply by ad-sharky** [reply]
Thanks for the quick response

**Reply by GoSpursGoNL** [reply]
Why does 'php artisan winter:update not update from 1.1.9 to 1.1.10?
Instead I get 'no new updates found'

**Reply by mjauvin** [reply]
Because we favor composer from now on and need to update the artisan command (and the backend update page) to use composer transparently underneath.

**Reply by GoSpursGoNL** [reply]
Thanks, clear. Then this page confuses me though: https://wintercms.com/docs/v1.2/docs/console/setup-maintenance

**Comment by arvislacis** [comment]
Not the real solution but maybe can help for you too:

For my projects I have configured endpoints like `/adm/composer-update` and `/adm/winter-up` - which are accessible only for logged in back-end users (admins) and when opened, triggers composer or artisan command line utilities to do composer package update or migrate plugins. I found this method useful and time saving if you need to maintain many environments (local, testing...) at the same time and don't want to login-in to server every time, especially if you are running projects with Docker.

**Reply by qroft** [reply]
I find this a super usefull solution to be honest as i have two main problems:
1. Composer is unknown land to me.
2. My shared webhost does not allow me to do some composer stuff, what is absolutely weird.

So: could you explain your workflow in detail?
I imagine that you created something in the php code section of a page, but what does that look like?
Thanks in advance!

**Reply by mjauvin** [reply]
@arvislacis Could you share your solution as a plugin, or in a trick at https://wintertricks.com/ ?

**Reply by arvislacis** [reply]
@qroft My solution still involves using `composer `and also `shell_exec()` (so it will probably not work for you, as your webhost doesn't allow to install/use composer). My solution only triggers composer cli by visting defined route/webhook but the composer part is pretty much the same.

Anyways, here is the route I have defined in my plugin:

```
 Route::get('/backend/composer-update', function () {
     ini_set('max_execution_time', 300);

     $response = [];
     $user = BackendAuth::getUser();

     if ($user && $user->is_superuser) {
         $basepath = base_path();
         $response['composerVersion'] = shell_exec("cd {$basepath} && composer -vvv about");
         $response['composerSelfUpdate'] = shell_exec("cd {$basepath} && composer self-update");
         $response['composerUpdate'] = shell_exec("cd {$basepath} && composer update");
         $response['winterUp'] = shell_exec("cd {$basepath} && php artisan winter:up");

         Log::info($user->login . ' (' . $user->id . ') triggered composer update');

         return response()->json($response, 200, [], JSON_PRETTY_PRINT);
     } else {
         return Redirect::to('/404');
     }
 })->middleware('web');
 ```

**Reply by bennothommo** [reply]
@qroft I am in no way targeting you with this, it's more of a general statement, however:

> Composer is unknown land to me.

I find this to be more and more an untenable position these days. Composer has been out for more than 12 years and nearly every modern PHP application, framework or utility uses it. Really, the only hold-out on this is Wordpress, and frankly, they are so far removed from normal PHP development anymore that they might as well be considered an entirely different ecosystem.

Nearly every language out there now has some sort of dependency management (JavaScript/Node has NPM, Ruby has Gems, etc...) and that is all Composer is. You will do yourself a great service to learn it and understand it.

Our intention is to implement Composer into our upgrade process in such a way that it is invisible to people using Winter CMS, but it still will be Composer, which means that people - especially those that wish to publish plugins and themes - will still need to know how it works. I respect that some people have hosts that don't allow them to use Composer on them directly - that's a fair reason to not use it there. But there is nothing stopping people from using it on their local copies, where no such limitation would exist, and then FTP'ing all files to their host.

There's always a way ;)

**Comment by arvislacis** [comment]
> @arvislacis Could you share your solution as a plugin, or in a trick at https://wintertricks.com/ ?

@mjauvin See my https://github.com/orgs/wintercms/discussions/279#discussioncomment-8854895 reply and also feel free to publish or modify it for usage under https://wintertricks.com Anyways, I see this solution a bit hacky for myself, and it still requires composer and shell_exec..., the only benefit is that you don't need to log in to your server or docker container, and maybe can configure this as webhook for some cron/background process.
